sqlmesh/core/snapshot/evaluator.py

```diff
@@ -175,7 +177,18 @@ def apply(query_or_df: QueryOrDF, index: int = 0) -> None:
                                 execute(exp.select(existing_limit.expression)).rows[0][0],
                             )
                     return query_or_df.head(limit) if hasattr(query_or_df, "head") else self.adapter._fetch_native_df(query_or_df.limit(limit))  # type: ignore
-
+                elif (
+                    self.adapter.SUPPORTS_INSERT_OVERWRITE
+                    and snapshot.is_incremental_by_time_range
+                    and df is not None
+                ):
+                    for next_df in queries_or_dfs:
```

Comment:
can we do instead 

reduce(lambda a, b: a.unionAll(b) if self.adapter.pyspark(a) else pd.concat([a, b]) queries_or_dfs

Reply:
Yeah I initial wrote that but the challenge is that `queries_or_dfs` has already yielded the initial DataFrame so you are just unioning the remaining Dataframes. After that you need to then union again the current dataframe into those unioned dataframes. Therefore I thought the current solution was better since we start with the dataframe we have already yielded.

Reply:
maybe refactor it a bit

if limit and limit > 0:
   return next(...).limit(...)
if  supports insert overwrite:
   apply(reduce(...),)
else:
  for enumerate ...

Reply:
This part is the problem: `if supports insert overwrite:`. It requires that we check the initial yielded result if it is a DataFrame since we only want to do the reduce concat logic if it is a DataFrame.

Reply:
@tobymao Updated: https://github.com/TobikoData/sqlmesh/pull/890/commits/916e0e11f6630545ec6848e49aab7a1c3fc9556a